protofsm/state_machine.go

```diff
@@ -0,0 +1,428 @@
+	close(s.quit)
+	s.wg.Wait()
+}
+
+// SendEvent sends a new event to the state machine.
+//
+// TODO(roasbeef): bool if processed?
+func (s *StateMachine[Event, Env]) SendEvent(event Event) {
+	select {
+	case s.events <- event:
+	case <-s.quit:
+		return
+	}
+}
+
+// CurrentState returns the current state of the state machine.
+func (s *StateMachine[Event, Env]) CurrentState() (State[Event, Env], error) {
+	query := stateQuery[Event, Env]{
+		CurrentState: make(chan State[Event, Env], 1),
+	}
+
+	if !fn.SendOrQuit(s.stateQuery, query, s.quit) {
+		return nil, fmt.Errorf("state machine is shutting down")
+	}
+
+	// TODO(roasbeef): channel sent on for each state transition?
+	// subscribers?
+
+	return fn.RecvOrTimeout(query.CurrentState, time.Second)
+}
+
+// StateSubscriber represents an active subscription to be notified of new
+// state transitions.
+type StateSubscriber[E any, F Environment] *fn.EventReceiver[State[E, F]]
+
+// RegisterStateEvents registers a new event listener that will be notified of
+// new state transitions.
+func (s *StateMachine[Event, Env]) RegisterStateEvents() StateSubscriber[Event, Env] {
+	subscriber := fn.NewEventReceiver[State[Event, Env]](10)
+
+	// TODO(roasbeef): instead give the state and the input event?
+
+	s.newStateEvents.RegisterSubscriber(subscriber)
+
+	return subscriber
+}
+
+// RemoveStateSub removes the target state subscriber from the set of active
+// subscribers.
+func (s *StateMachine[Event, Env]) RemoveStateSub(sub StateSubscriber[Event, Env]) {
+	s.newStateEvents.RemoveSubscriber(sub)
+}
+
+// executeDaemonEvent executes a daemon event, which is a special type of event
+// that can be emitted as part of the state transition function of the state
+// machine. An error is returned if the type of event is unknown.
+func (s *StateMachine[Event, Env]) executeDaemonEvent(event DaemonEvent) error {
+	switch daemonEvent := event.(type) {
+
+	// This is a send message event, so we'll send the event, and also mind
+	// any preconditions as well as post-send events.
+	case *SendMsgEvent[Event]:
+		sendAndCleanUp := func() error {
+			err := s.daemon.SendMessages(
+				daemonEvent.TargetPeer, daemonEvent.Msgs,
+			)
+			if err != nil {
+				return fmt.Errorf("unable to send msgs: %w", err)
+			}
+
+			// If a post-send event was specified, then we'll
+			// funnel that back into the main state machine now as
+			// well.
+			daemonEvent.PostSendEvent.WhenSome(func(event Event) {
+				s.wg.Add(1)
+				go func() {
+					defer s.wg.Done()
+
+					s.SendEvent(event)
+				}()
+			})
+
+			return nil
+		}
+
+		// If this doesn't have a SendWhen predicate, then we can just
+		// send it off right away.
+		if !daemonEvent.SendWhen.IsSome() {
+			if err := sendAndCleanUp(); err != nil {
+				return err
+			}
+		}
+
+		// Otherwise, this has a SendWhen predicate, so we'll need
+		// launch a goroutine to poll the SendWhen, then send only once
+		// the predicate is true.
```

Comment:
Yeah agree the polling (as you mentioned) has the downside on lagging the first instance where the new state/predicate is present, and in the case of something that can revert (flip on, then off) it may miss the opportunity all together. 

> If it's not prohibitively difficult I'd suggest we find a way to make SendWhen's predicate pure, and actually feed it the state changes it's monitoring rather than time based polling of state, but I do recognize that may not be an easy lift.

Hmm, yeah I'm not sure _what_ we would pass in here. With the code as is in the RBF closer PR, it could pass in the `ChanStateObserver` (or w/e it's called now), but then that would require the FSM executor to be able to extract attributes from the env, whereas rn it's based on an interface.

Reply:
Yeah it's very unclear what you would want to send I agree. Ideally it's the state we are polling that says "coast is clear" and send in a new copy of it on each change. Generally I think of the relevant state here as channel state (the state of balances and commitment transactions) and link state (the state associated with where in the protocol trace we are with our peer). LightningChannel captures the channel state well but we don't really have a good tracker of link state since our link is busted up across multiple different data structures with the chan closers and funding manager. Making the desired change here may require us finishing the ChannelLifecycle refactor. Maybe we can try out a polling answer and just see what happens? It can mean untimely shutdown responses though.